File: server/entitlement/service.ts
import type { VaultContext } from '../../lib/domain/identity';
import {
  authorizeEntitlementState,
  authorizeOfflineLease,
  evaluateSubscriptionFacts,
  planEntitlementProjection,
  planOfflineLease,
  type EntitlementProjectionRecord,
  type OfflineLeaseRecord,
} from './core';
import type { EntitlementDependencies } from './ports';
import {
  paidPersonalVaultLimits,
  type EntitlementCapability,
  type EntitlementDecision,
  type EntitlementDenialReason,
  type EntitlementLimitDecision,
  type EntitlementPort,
  type OfflineLeaseCommandResult,
  type OfflineLease,
  type OfflineLeasePolicy,
} from './public';

type ProjectionResult =
  | { readonly kind: 'ready'; readonly record: EntitlementProjectionRecord }
  | { readonly kind: 'denied'; readonly reason: EntitlementDenialReason };

export function createEntitlementPort(
  dependencies: EntitlementDependencies,
  offlineLeasePolicy: OfflineLeasePolicy,
): EntitlementPort {
  return {
    async authorizeCapability(
      context,
      capability,
      checkedAt,
    ): Promise<EntitlementDecision> {
      const owner = await verifyOwner(dependencies, context);
      if (owner !== 'owned') return denied(capability, owner);
      const projection = await refreshProjection(
        dependencies,
        context,
        checkedAt,
      );
      if (projection.kind === 'denied') {
        return isRecoveryCapability(capability) &&
          projection.reason !== 'invalid-input'
          ? recoveryAllowed(capability)
          : denied(capability, projection.reason);
      }
      return authorizeEntitlementState(
        projection.record.state,
        capability,
        checkedAt,
      );
    },

    async readLimits(context, checkedAt): Promise<EntitlementLimitDecision> {
      const owner = await verifyOwner(dependencies, context);
      if (owner !== 'owned') return { kind: 'denied', reason: owner };
      const projection = await refreshProjection(
        dependencies,
        context,
        checkedAt,
      );
      if (projection.kind === 'denied') return projection;
      const decision = authorizeEntitlementState(
        projection.record.state,
        'notes-read',
        checkedAt,
      );
      return decision.kind === 'allowed' && decision.validUntil !== null
        ? {
            kind: 'available',
            limits: paidPersonalVaultLimits,
            validUntil: decision.validUntil,
          }
        : {
            kind: 'denied',
            reason:
              decision.kind === 'denied'
                ? decision.reason
                : 'entitlement-unavailable',
          };
    },

    async issueOfflineLease(
      context,
      input,
    ): Promise<OfflineLeaseCommandResult> {
      const owner = await verifyOwner(dependencies, context);
      if (owner !== 'owned') return { kind: 'denied', reason: owner };
      if (offlineLeasePolicy.kind === 'undecided') {
        return { kind: 'denied', reason: 'lease-policy-undecided' };
      }
      const projection = await refreshProjection(
        dependencies,
        context,
        input.issuedAt,
      );
      if (projection.kind === 'denied') return projection;
      const plan = planOfflineLease(
        context,
        projection.record,
        offlineLeasePolicy,
        input,
      );
      if (plan.kind === 'denied') return plan;
      try {
        const created = await dependencies.repository.createOfflineLease({
          expectedProjectionVersion: projection.record.version,
          lease: plan.lease,
        });
        switch (created.kind) {
          case 'issued':
            return { kind: 'issued', lease: toOfflineLease(plan.lease) };
          case 'replayed':
            return { kind: 'replayed', lease: toOfflineLease(created.lease) };
          case 'identifier-conflict':
            return { kind: 'denied', reason: 'identifier-conflict' };
          case 'projection-conflict':
            return { kind: 'denied', reason: 'projection-conflict' };
        }
      } catch {
        return { kind: 'denied', reason: 'entitlement-unavailable' };
      }
    },

    async authorizeOfflineCapability(
      context,
      capability,
      leaseId,
      checkedAt,
    ): Promise<EntitlementDecision> {
      const owner = await verifyOwner(dependencies, context);
      if (owner !== 'owned') return denied(capability, owner);
      try {
        const lease = await dependencies.repository.findOfflineLease(
          context,
          leaseId,
        );
        return lease === undefined
          ? denied(capability, 'lease-not-found')
          : authorizeOfflineLease(lease, context, capability, checkedAt);
      } catch {
        return denied(capability, 'entitlement-unavailable');
      }
    },
  };
}

async function refreshProjection(
  dependencies: EntitlementDependencies,
  context: VaultContext,
  checkedAt: number,
): Promise<ProjectionResult> {
  if (!Number.isSafeInteger(checkedAt) || checkedAt < 0) {
    return { kind: 'denied', reason: 'invalid-input' };
  }
  let facts;
  try {
    facts = await dependencies.billing.readSubscription(context);
  } catch {
    return { kind: 'denied', reason: 'billing-unavailable' };
  }
  if (facts === undefined) {
    return { kind: 'denied', reason: 'subscription-required' };
  }
  const evaluation = evaluateSubscriptionFacts(facts, checkedAt);
  if (evaluation.kind === 'invalid') {
    return { kind: 'denied', reason: 'billing-unavailable' };
  }
  for (let attempt = 0; attempt < 2; attempt += 1) {
    let current;
    try {
      current = await dependencies.repository.findProjection(context);
    } catch {
      return { kind: 'denied', reason: 'entitlement-unavailable' };
    }
    const plan = planEntitlementProjection(
      context,
      facts,
      evaluation.state,
      checkedAt,
      current,
    );
    switch (plan.kind) {
      case 'invalid':
        return { kind: 'denied', reason: 'billing-unavailable' };
      case 'stale':
        return { kind: 'denied', reason: 'projection-conflict' };
      case 'current':
        return { kind: 'ready', record: plan.record };
      case 'commit':
        try {
          const committed = await dependencies.repository.commitProjection({
            expectedVersion: current?.version,
            record: plan.record,
            revokeActiveLeasesAt:
              plan.record.state.kind === 'locked' ? checkedAt : null,
          });
          if (committed.kind === 'applied') {
            return { kind: 'ready', record: plan.record };
          }
        } catch {
          return { kind: 'denied', reason: 'entitlement-unavailable' };
        }
        break;
    }
  }
  return { kind: 'denied', reason: 'projection-conflict' };
}

async function verifyOwner(
  dependencies: EntitlementDependencies,
  context: VaultContext,
): Promise<'owned' | 'owner-mismatch' | 'entitlement-unavailable'> {
  try {
    return (await dependencies.ownership.owns(context))
      ? 'owned'
      : 'owner-mismatch';
  } catch {
    return 'entitlement-unavailable';
  }
}

function isRecoveryCapability(capability: EntitlementCapability): boolean {
  return (
    capability === 'billing-recovery' ||
    capability === 'subscription-cancel' ||
    capability === 'account-delete' ||
    capability === 'support'
  );
}

function recoveryAllowed(
  capability: EntitlementCapability,
): EntitlementDecision {
  return {
    kind: 'allowed',
    capability,
    basis: 'recovery',
    validUntil: null,
  };
}

function toOfflineLease(record: OfflineLeaseRecord): OfflineLease {
  return {
    leaseId: record.leaseId,
    context: record.context,
    basis: record.basis,
    issuedAt: record.issuedAt,
    expiresAt: record.expiresAt,
    revokedAt: record.revokedAt,
  };
}

function denied(
  capability: EntitlementCapability,
  reason: EntitlementDenialReason,
): EntitlementDecision {
  return { kind: 'denied', capability, reason };
}

File: backend/internal/entitlement/service.go
package entitlement

import (
	"context"
	"errors"

	"github.com/fukamu/notes/backend/internal/identity"
)

var ErrInvalidServiceConfiguration = errors.New("invalid entitlement service configuration")

type Service struct {
	billing    BillingPort
	ownership  OwnershipPort
	repository Repository
	policy     OfflineLeasePolicy
}

func NewService(
	billingPort BillingPort,
	ownership OwnershipPort,
	repository Repository,
	policy OfflineLeasePolicy,
) (*Service, error) {
	if billingPort == nil || ownership == nil || repository == nil || !validPolicy(policy) {
		return nil, ErrInvalidServiceConfiguration
	}
	return &Service{billing: billingPort, ownership: ownership, repository: repository, policy: policy}, nil
}

func (service *Service) AuthorizeCapability(
	ctx context.Context,
	vaultContext identity.VaultContext,
	capability Capability,
	checkedAt int64,
) Decision {
	if service == nil || !validCapability(capability) || !ValidVaultContext(vaultContext) || !validTimestamp(checkedAt) {
		return denied(capability, DenialInvalidInput)
	}
	owner := service.verifyOwner(ctx, vaultContext)
	if owner != "" {
		return denied(capability, owner)
	}
	projection, reason := service.refreshProjection(ctx, vaultContext, checkedAt)
	if reason != "" {
		if isRecoveryCapability(capability) && reason != DenialInvalidInput {
			return Decision{Kind: DecisionAllowed, Capability: capability, Basis: BasisRecovery}
		}
		return denied(capability, reason)
	}
	return AuthorizeState(projection.State, capability, checkedAt)
}

func (service *Service) ReadLimits(
	ctx context.Context,
	vaultContext identity.VaultContext,
	checkedAt int64,
) LimitDecision {
	if service == nil || !ValidVaultContext(vaultContext) || !validTimestamp(checkedAt) {
		return LimitDecision{Kind: LimitsDenied, Reason: DenialInvalidInput}
	}
	if reason := service.verifyOwner(ctx, vaultContext); reason != "" {
		return LimitDecision{Kind: LimitsDenied, Reason: reason}
	}
	projection, reason := service.refreshProjection(ctx, vaultContext, checkedAt)
	if reason != "" {
		return LimitDecision{Kind: LimitsDenied, Reason: reason}
	}
	decision := AuthorizeState(projection.State, CapabilityNotesRead, checkedAt)
	if decision.Kind != DecisionAllowed || decision.ValidUntil == nil {
		if decision.Reason == "" {
			return LimitDecision{Kind: LimitsDenied, Reason: DenialEntitlementUnavailable}
		}
		return LimitDecision{Kind: LimitsDenied, Reason: decision.Reason}
	}
	return LimitDecision{Kind: LimitsAvailable, Limits: PaidPersonalVaultLimits(), ValidUntil: *decision.ValidUntil}
}

func (service *Service) IssueOfflineLease(
	ctx context.Context,
	vaultContext identity.VaultContext,
	leaseID OfflineLeaseID,
	issuedAt int64,
) LeaseCommandResult {
	if service == nil || !ValidVaultContext(vaultContext) || !validTimestamp(issuedAt) {
		return LeaseCommandResult{Kind: LeaseDenied, Reason: DenialInvalidInput}
	}
	if _, err := ParseOfflineLeaseID(string(leaseID)); err != nil {
		return LeaseCommandResult{Kind: LeaseDenied, Reason: DenialInvalidInput}
	}
	if reason := service.verifyOwner(ctx, vaultContext); reason != "" {
		return LeaseCommandResult{Kind: LeaseDenied, Reason: reason}
	}
	if service.policy.Kind == OfflineLeaseUndecided {
		return LeaseCommandResult{Kind: LeaseDenied, Reason: DenialLeasePolicyUndecided}
	}
	projection, reason := service.refreshProjection(ctx, vaultContext, issuedAt)
	if reason != "" {
		return LeaseCommandResult{Kind: LeaseDenied, Reason: reason}
	}
	plan := PlanOfflineLease(vaultContext, projection, service.policy, leaseID, issuedAt)
	if plan.Kind == OfflineLeaseDeny {
		return LeaseCommandResult{Kind: LeaseDenied, Reason: plan.Reason}
	}
	created, err := service.repository.CreateOfflineLease(ctx, projection.Version, plan.Lease)
	if err != nil {
		return LeaseCommandResult{Kind: LeaseDenied, Reason: DenialEntitlementUnavailable}
	}
	switch created.Kind {
	case LeaseCreateIssued:
		lease := publicLease(plan.Lease)
		return LeaseCommandResult{Kind: LeaseIssued, Lease: &lease}
	case LeaseCreateReplayed:
		if created.Lease == nil || !ValidOfflineLeaseRecord(*created.Lease) {
			return LeaseCommandResult{Kind: LeaseDenied, Reason: DenialEntitlementUnavailable}
		}
		lease := publicLease(*created.Lease)
		return LeaseCommandResult{Kind: LeaseReplayed, Lease: &lease}
	case LeaseCreateIdentifierConflict:
		return LeaseCommandResult{Kind: LeaseDenied, Reason: DenialIdentifierConflict}
	case LeaseCreateProjectionConflict:
		return LeaseCommandResult{Kind: LeaseDenied, Reason: DenialProjectionConflict}
	default:
		return LeaseCommandResult{Kind: LeaseDenied, Reason: DenialEntitlementUnavailable}
	}
}

func (service *Service) AuthorizeOfflineCapability(
	ctx context.Context,
	vaultContext identity.VaultContext,
	capability Capability,
	leaseID OfflineLeaseID,
	checkedAt int64,
) Decision {
	if service == nil || !validCapability(capability) || !ValidVaultContext(vaultContext) || !validTimestamp(checkedAt) {
		return denied(capability, DenialInvalidInput)
	}
	if _, err := ParseOfflineLeaseID(string(leaseID)); err != nil {
		return denied(capability, DenialInvalidInput)
	}
	if reason := service.verifyOwner(ctx, vaultContext); reason != "" {
		return denied(capability, reason)
	}
	lease, err := service.repository.FindOfflineLease(ctx, vaultContext, leaseID)
	if err != nil {
		return denied(capability, DenialEntitlementUnavailable)
	}
	if lease == nil {
		return denied(capability, DenialLeaseNotFound)
	}
	if !ValidOfflineLeaseRecord(*lease) {
		return denied(capability, DenialEntitlementUnavailable)
	}
	return AuthorizeOfflineLease(lease.OfflineLease, vaultContext, capability, checkedAt)
}

func (service *Service) refreshProjection(
	ctx context.Context,
	vaultContext identity.VaultContext,
	checkedAt int64,
) (ProjectionRecord, DenialReason) {
	facts, err := service.billing.ReadSubscription(ctx, vaultContext)
	if err != nil {
		return ProjectionRecord{}, DenialBillingUnavailable
	}
	if facts == nil {
		return ProjectionRecord{}, DenialSubscriptionRequired
	}
	evaluation := EvaluateSubscriptionFacts(*facts, checkedAt)
	if evaluation.Kind != EvaluationEvaluated {
		return ProjectionRecord{}, DenialBillingUnavailable
	}
	for attempt := 0; attempt < 2; attempt++ {
		current, findErr := service.repository.FindProjection(ctx, vaultContext)
		if findErr != nil {
			return ProjectionRecord{}, DenialEntitlementUnavailable
		}
		plan := PlanProjection(vaultContext, *facts, evaluation.State, checkedAt, current)
		switch plan.Kind {
		case ProjectionInvalid:
			return ProjectionRecord{}, DenialBillingUnavailable
		case ProjectionStale:
			return ProjectionRecord{}, DenialProjectionConflict
		case ProjectionCurrent:
			return plan.Record, ""
		case ProjectionCommit:
			var expected *ProjectionVersion
			if current != nil {
				value := current.Version
				expected = &value
			}
			var revokeAt *int64
			if plan.Record.State.Kind == StateLocked {
				revokeAt = pointer(checkedAt)
			}
			committed, commitErr := service.repository.CommitProjection(ctx, expected, plan.Record, revokeAt)
			if commitErr != nil {
				return ProjectionRecord{}, DenialEntitlementUnavailable
			}
			if committed == ProjectionApplied {
				return plan.Record, ""
			}
		default:
			return ProjectionRecord{}, DenialEntitlementUnavailable
		}
	}
	return ProjectionRecord{}, DenialProjectionConflict
}

func (service *Service) verifyOwner(ctx context.Context, vaultContext identity.VaultContext) DenialReason {
	owned, err := service.ownership.Owns(ctx, vaultContext)
	if err != nil {
		return DenialEntitlementUnavailable
	}
	if !owned {
		return DenialOwnerMismatch
	}
	return ""
}

func isRecoveryCapability(capability Capability) bool {
	return capability == CapabilityBillingRecovery || capability == CapabilitySubscriptionCancel ||
		capability == CapabilityAccountDelete || capability == CapabilitySupport
}

func publicLease(record OfflineLeaseRecord) OfflineLease {
	return OfflineLease{
		LeaseID: record.LeaseID, Context: record.Context, Basis: record.Basis,
		IssuedAt: record.IssuedAt, ExpiresAt: record.ExpiresAt, RevokedAt: cloneTimestamp(record.RevokedAt),
	}
}
